### PR TITLE
Mask RCNN and more

### PR DESCRIPTION
**Pull request by 0x00b1**
In additional to the numerous fixes (and tests), a number of important features are also included:

- a Data Science Bowl (DSB) 2018 dataset
- refactored loss layers
- an entirely rewritten ObjectDetectionGenerator
- a Mask RCNN implementation

### CODE REVIEW

File: keras_rcnn/preprocessing/_object_detection.py [hunk 1]
```diff
@@ -38,19 +62,33 @@ def __init__(
         self.generator = generator
 
         if self.color_mode == "grayscale":
-            if self.data_format == "channels_last":
-                self.image_shape = target_size + (1,)
+            if self.data_format == "channels_first":
+                self.image_shape = (*target_size, 1)
```
Comment (jhung0) [inline keras_rcnn/preprocessing/_object_detection.py]: Why not use self.channels from line 48?

**Comment by theone4ever** [comment]
It seems the bounding  boxes are rotated when using ObjectDetectionGenerator with dataset DSB2018. So in _object_detection.py line 189, instead of ```target_bounding_box = [
                    minimum_r,
                    minimum_c,
                    maximum_r,
                    maximum_c
                ]```, it should be ```target_bounding_box = [
                    minimum_c,
                    minimum_r,
                    maximum_c,
                    maximum_r
                ]```

**Comment by 0x00b1** [comment]
unknown

File: keras_rcnn/models/_rcnn.py [hunk 2]
```diff
@@ -12,78 +12,90 @@
...
+        output_deltas = keras.layers.Conv2D(9 * 4, (1, 1), activation="linear", kernel_initializer="zero", name="deltas")(convolution_3x3)
 
-        output_features = keras.layers.Conv2D(512, name='block4_conv1', **options)(output_features)
-        output_features = keras.layers.Conv2D(512, name='block4_conv2', **options)(output_features)
-        output_features = keras.layers.Conv2D(512, name='block4_conv3', **options)(output_features)
-        output_features = keras.layers.Conv2D(512, name='block4_conv4', **options)(output_features)
+        output_scores = keras.layers.Conv2D(9 * 1, (1, 1), activation="sigmoid", kernel_initializer="uniform", name="scores")(convolution_3x3)
 
-        output_features = keras.layers.MaxPooling2D(strides=(2, 2), name='block4_pool')(output_features)
+        target_anchors, target_proposal_categories, target_proposal_bounding_boxes = keras_rcnn.layers.AnchorTarget()([output_scores, target_bounding_boxes, target_metadata])
```
Comment (jhung0) [inline keras_rcnn/models/_rcnn.py]: the order should be bounding_boxes, scores

File: keras_rcnn/models/_rcnn.py [hunk 3]
```diff
@@ -12,78 +12,90 @@
...
 
-        output_features = keras.layers.Conv2D(512, name='block5_conv1', **options)(output_features)
-        output_features = keras.layers.Conv2D(512, name='block5_conv2', **options)(output_features)
-        output_features = keras.layers.Conv2D(512, name='block5_conv3', **options)(output_features)
+        output_deltas, output_scores = keras_rcnn.layers.RPN()([target_proposal_bounding_boxes, target_proposal_categories, output_deltas, output_scores])
 
-        convolution_3x3 = keras.layers.Conv2D(512, name="convolution_3x3", **options)(output_features)
+        output_proposal_categories = keras_rcnn.layers.ObjectProposal()([target_metadata, output_deltas, output_scores, target_anchors])
```
Comment (jhung0) [inline keras_rcnn/models/_rcnn.py]: reorder: target anchors first, metadata last

File: keras_rcnn/models/_rcnn.py [hunk 4]
```diff
@@ -12,78 +12,90 @@
...
 
-        output_deltas = keras.layers.Conv2D(9 * 4, (1, 1), activation="linear", kernel_initializer="zero", name="deltas")(convolution_3x3)
-        output_scores = keras.layers.Conv2D(9 * 1, (1, 1), activation="sigmoid", kernel_initializer="uniform", name="scores")(convolution_3x3)
+        output_proposal_categories, target_proposal_categories, target_proposal_bounding_boxes = keras_rcnn.layers.ProposalTarget()([output_proposal_categories, target_categories, target_bounding_boxes])
```
Comment (jhung0) [inline keras_rcnn/models/_rcnn.py]: reorder input and output: target bounding boxes, target categories, output proposal categories

File: keras_rcnn/models/_rcnn.py [hunk 5]
```diff
@@ -12,78 +12,90 @@
...
 
-        target_anchors, target_proposal_labels, target_proposals = keras_rcnn.layers.AnchorTarget()([output_scores, target_bounding_boxes, target_metadata])
+        output_features = keras_rcnn.layers.RegionOfInterest((14, 14))([output_features, output_proposal_categories, target_metadata])
```
Comment (jhung0) [inline keras_rcnn/models/_rcnn.py]: should the 14 be in a config?

File: keras_rcnn/models/_rcnn.py (hunk 3 above)
Comment (jhung0) [reply keras_rcnn/models/_rcnn.py]: Also I was confused at first what output proposal categories was (at first I thought they were something to do with classes but they don't)

File: keras_rcnn/models/_rcnn.py (hunk 5 above)
Comment (0x00b1) [reply keras_rcnn/models/_rcnn.py]: Good question. 

It’s probably something we wouldn’t encourage users to configure since it’ll usually lead to worse results, but it’s also something users will occasionally need to configure. If we do want to add it, I guess we should make it a parameter of the RCNN model. What do you think?

File: keras_rcnn/models/_rcnn.py (hunk 3 above)
Comment (0x00b1) [reply keras_rcnn/models/_rcnn.py]: What does it return? The proposed bounding boxes? Objectness scores? Both?

Comment (jhung0) [reply keras_rcnn/models/_rcnn.py]: no, just proposed bounding boxes

File: keras_rcnn/models/_rcnn.py (hunk 5 above)
Comment (jhung0) [reply keras_rcnn/models/_rcnn.py]: Yeah I'd agree

File: keras_rcnn/layers/object_detection/_anchor_target.py [hunk 6]
```diff
@@ -59,11 +59,11 @@ def build(self, input_shape):
 
     # TODO: should AnchorTarget only be enabled during training
     def call(self, inputs, **kwargs):
-        scores, gt_boxes, metadata = inputs
+        target_bounding_boxes, metadata, scores = inputs
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/_anchor_target.py]: `target_metadata`

File: keras_rcnn/layers/object_detection/_object_proposal.py [hunk 7]
```diff
@@ -38,7 +38,7 @@ def call(self, inputs, **kwargs):
         """
         `image_shape_and_scale` has the shape [width, height, scale]
         """
-        image_shape_and_scale, deltas, scores, anchors = inputs
+        anchors, image_shape_and_scale, deltas, scores = inputs
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/_object_proposal.py]: rename `image_shape_and_scale` to `target_metadata`

File: keras_rcnn/layers/object_detection/_proposal_target.py [hunk 8]
```diff
@@ -72,7 +72,7 @@ def call(self, inputs, training=None):
 
         # labels (class1, class2, ... , num_classes)
         # Include ground-truth boxes in the set of candidate rois
-        proposals, labels, bounding_boxes = inputs
+        bounding_boxes, labels, proposals = inputs
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/_proposal_target.py]: rename `bounding_boxes` to `target_bounding_boxes`

Comment (0x00b1) [inline keras_rcnn/layers/object_detection/_proposal_target.py]: rename `labels` to `target_categories`

File: keras_rcnn/models/_rcnn.py [hunk 9]
```diff
@@ -12,78 +12,90 @@
...
+        inputs = [
+            target_bounding_boxes,
+            target_categories,
+            target_image,
+            target_masks,
+            target_metadata
+        ]
 
-        output_features = keras.layers.MaxPooling2D(strides=(2, 2), name='block2_pool')(output_features)
+        output_features = keras.layers.Conv2D(64, **options)(target_image)
```
Comment (0x00b1) [inline keras_rcnn/models/_rcnn.py]: Replace with a standard backbone architecture.

File: keras_rcnn/layers/_object_detection.py [hunk 10]
```diff
@@ -34,7 +34,7 @@ def call(self, x, training=None, **kwargs):
         """
 
         def detections(num_output):
-            proposals, deltas, scores, metadata = x[0], x[1], x[2], x[3]
+            metadata, deltas, proposals, scores = x[0], x[1], x[2], x[3]
```
[the remark this answers is not in the document]
Comment (drwaltman) [reply keras_rcnn/layers/_object_detection.py]: [#191](https://github.com/broadinstitute/keras-rcnn/pull/191)
